src/mercury-common/mercury/common/configuration.py

```diff
@@ -79,3 +84,368 @@ def get_configuration(filename):
+
+        .. code-block::
+
+            {'one': {'two': 'three'}}
+
+            >> d['one']['two']
+            'three'
+            >> get_by_namespace('one.two')
+            'three'
+        :param dictionary: The dictionary we are searching
+        :param namespace: Dot notated namespace specifier
+        :type namespace: str
+        :return: The value the namespace resolves to or None
+        """
+
+        # creating copy of reference for readability
+        value = dictionary
+        for k in namespace.split('.'):
+            try:
+                value = value[k]
+            except KeyError:
+                return
+        return value
+
+    @staticmethod
+    def set_by_namespace(dictionary, namespace, value):
+        """
+        Like get_by_namespace, but set a value rather than returning it
+        :param dictionary: The dictionary we are modifying
+        :param namespace:
+        :param value:
+        :return:
+        """
+        expanded = namespace.split('.')
+        for k in expanded[:-1]:
+            dictionary = dictionary[k]
+        dictionary[expanded[-1]] = value
+
+    @staticmethod
+    def convert_argparse_destination(argument_name):
+        """ Converts underscores to dashes, for a deterministic mechanism
+         to name (and recall) argparse destinations"""
+        return argument_name.lstrip('--').replace('-', '_')
+
+    def parse_args(self):
+        """ Add option arguments and parse_args """
+        for opt in self.options:
+            if not opt['added_to_parser']:
+                self.argument_parser.add_argument(
+                    opt['cli_argument'],
+                    dest=self.convert_argparse_destination(opt['cli_argument']),
+                    help=opt['help'],
+                    *opt['args'],
+                    **opt['kwargs'])
+                # added_to_parser is here for instances where
+                # populate_namespace is called more than once per class
+                # instance. Such a thing happens in our test cases, and thus
+                # could happen during implementation
+                opt['added_to_parser'] = True
+
+        self.argparse_namespace = self.argument_parser.parse_args()
+
+    @staticmethod
+    def format_type(value, special_type):
+        """
+        Since argparse (by default) and os.environ can only be strings, we allow
+        options to specify a type
+        :param value: The value we are transforming
+        :param special_type: The 'type' we are transforming to
+        :type special_type: list, int, float, bool
+        :return: None
+        """
+
+        if value is None or special_type is None:
+            # Either value or special type are not set
+            # Check this here to prevent more branching up stream
+            return value
+
+        if type(special_type) != type:
+            raise MercuryConfigurationError(
+                f'{special_type} must be a type, not an object'
```

Comment:
First time I see f strings!

Reply:
it's a Py3 thing - https://www.python.org/dev/peps/pep-0498/; I don't think that got ported to Py2

Reply:
@xzased I started mercury using C style formatting % .. I figure I would use f-strings to enforce a python 3.6 requirement :)